File: app/src/main/java/com/amazon/tv/leanbacklauncher/apps/SettingsAdapter.java
package com.amazon.tv.leanbacklauncher.apps;

import android.content.Context;
import android.content.pm.PackageManager.NameNotFoundException;
import android.content.res.Resources;
import android.graphics.drawable.Drawable;
import android.os.Handler;
import android.os.Message;

import com.amazon.tv.firetv.leanbacklauncher.apps.AppCategory;
import com.amazon.tv.leanbacklauncher.R;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;

public class SettingsAdapter extends AppsAdapter {
    private final ConnectivityListener mConnectivityListener;
    private final Handler mHandler = new NetworkUpdateHandler(this);
    private Resources mNetResources;
    private boolean mNetResourcesSet = false;
    private static String TAG = "SettingsAdapter";

    private static class NetworkUpdateHandler extends Handler {
        private final WeakReference<SettingsAdapter> mAdapterRef;

        public NetworkUpdateHandler(SettingsAdapter adapter) {
            this.mAdapterRef = new WeakReference(adapter);
        }

        public void handleMessage(Message msg) {
            SettingsAdapter adapter = this.mAdapterRef.get();
            if (adapter != null) {
                switch (msg.what) {
                    case 1:
                        int index = adapter.updateNetwork();
                        if (index >= 0) {
                            adapter.notifyItemChanged(index);
                            return;
                        }
                        return;
                    default:
                        return;
                }
            }
        }
    }

    private class SettingsComparator implements Comparator<LaunchPoint> {
        private SettingsComparator() {
        }

        public int compare(LaunchPoint lhs, LaunchPoint rhs) {
            if (lhs.getPriority() != rhs.getPriority()) {
                return rhs.getPriority() - lhs.getPriority();
            }
            if (lhs.getTitle() == null) {
                return -1;
            }
            if (rhs.getTitle() == null) {
                return 1;
            }
            return lhs.getTitle().compareToIgnoreCase(rhs.getTitle());
        }
    }

    public SettingsAdapter(Context context, ConnectivityListener listener) {
        super(context, null, AppCategory.SETTINGS);
        this.mConnectivityListener = listener;
    }

    protected final void onPostRefresh() {
        this.mNetResourcesSet = false;
        updateNetwork();
    }

    public void onConnectivityChange() {
        this.mHandler.sendEmptyMessage(1);
    }

    private int updateNetwork() {
        for (int i = 0; i < this.mLaunchPoints.size(); i++) {
            LaunchPoint launchPoint = this.mLaunchPoints.get(i);
            if (launchPoint.getSettingsType() == 0) {
                setNetwork(this.mContext.getResources(), launchPoint);
                return i;
            }
        }
        return -1;
    }

    private LaunchPoint setNetwork(Resources res, LaunchPoint launchPoint) {
        String str = null;
        if (!this.mNetResourcesSet) {
            setNetworkResources(launchPoint);
        }
        ConnectivityListener.ConnectivityStatus connectivityStatus = this.mConnectivityListener.getConnectivityStatus();
        int titleId = 0;
        String title = null;
        boolean hasNetworkName = false;
        launchPoint.setIconDrawable(null);
        switch (connectivityStatus.mNetworkType) {
            case 1:
                titleId = R.string.settings_network;
                launchPoint.setIconDrawable(getNetResourceDrawable(launchPoint, "network_state_disconnected"));
                break;
            case 3:
            case 5:
                title = connectivityStatus.mWifiSsid;
                hasNetworkName = true;
                switch (connectivityStatus.mWifiSignalStrength) {
                    case 0:
                        launchPoint.setIconDrawable(getNetResourceDrawable(launchPoint, "network_state_wifi_0"));
                        break;
                    case 1:
                        launchPoint.setIconDrawable(getNetResourceDrawable(launchPoint, "network_state_wifi_1"));
                        break;
                    case 2:
                        launchPoint.setIconDrawable(getNetResourceDrawable(launchPoint, "network_state_wifi_2"));
                        break;
                    case 3:
                        launchPoint.setIconDrawable(getNetResourceDrawable(launchPoint, "network_state_wifi_3"));
                        break;
                    case 4:
                        launchPoint.setIconDrawable(getNetResourceDrawable(launchPoint, "network_state_wifi_4"));
                        break;
                    default:
                        break;
                }
                break;
            case 7:
                launchPoint.setIconDrawable(getNetResourceDrawable(launchPoint, "network_state_ethernet"));
                titleId = R.string.settings_network;
                break;
            case 9:
                title = connectivityStatus.mWifiSsid;
                hasNetworkName = true;
                launchPoint.setIconDrawable(getNetResourceDrawable(launchPoint, "network_state_wifi_no_internet"));
                break;
            case 11:
                titleId = R.string.settings_network;
                launchPoint.setIconDrawable(getNetResourceDrawable(launchPoint, "network_state_ethernet_no_internet"));
                break;
            case 13:
                title = connectivityStatus.mMobileNetworkName;
                hasNetworkName = true;
                switch (connectivityStatus.mMobileSignalStrength) {
                    case 0:
                        launchPoint.setIconDrawable(getNetResourceDrawable(launchPoint, "network_state_cellular_0"));
                        break;
                    case 1:
                        launchPoint.setIconDrawable(getNetResourceDrawable(launchPoint, "network_state_cellular_1"));
                        break;
                    case 2:
                        launchPoint.setIconDrawable(getNetResourceDrawable(launchPoint, "network_state_cellular_2"));
                        break;
                    case 3:
                        launchPoint.setIconDrawable(getNetResourceDrawable(launchPoint, "network_state_cellular_3"));
                        break;
                    case 4:
                        launchPoint.setIconDrawable(getNetResourceDrawable(launchPoint, "network_state_cellular_4"));
                        break;
                    default:
                        break;
                }
                break;
            case 15:
                title = connectivityStatus.mMobileNetworkName;
                hasNetworkName = true;
                switch (connectivityStatus.mMobileSignalStrength) {
                    case 0:
                        launchPoint.setIconDrawable(getNetResourceDrawable(launchPoint, "network_state_cellular_no_internet_0"));
                        break;
                    case 1:
                        launchPoint.setIconDrawable(getNetResourceDrawable(launchPoint, "network_state_cellular_no_internet_1"));
                        break;
                    case 2:
                        launchPoint.setIconDrawable(getNetResourceDrawable(launchPoint, "network_state_cellular_no_internet_2"));
                        break;
                    case 3:
                        launchPoint.setIconDrawable(getNetResourceDrawable(launchPoint, "network_state_cellular_no_internet_3"));
                        break;
                    case 4:
                        launchPoint.setIconDrawable(getNetResourceDrawable(launchPoint, "network_state_cellular_no_internet_4"));
                        break;
                    default:
                        break;
                }
        }
        if (launchPoint.getIconDrawable() == null) {
            launchPoint.setIconDrawable(getNetResourceDrawable(launchPoint, "network_state_wifi_0"));
        }
        if (titleId != 0) {
            title = res.getString(titleId);
        }
        if (title == null) {
            title = res.getString(R.string.settings_network);
        }
        launchPoint.setTitle(title);
        if (hasNetworkName) {
            str = res.getString(R.string.settings_network);
            launchPoint.setContentDescription(str);
        }
        return launchPoint;
    }

    public void onRankerReady() {
    }

    protected final void sortLaunchPoints(ArrayList<LaunchPoint> launchPoints) {
        Collections.sort(launchPoints, new SettingsComparator());
    }

    private void setNetworkResources(LaunchPoint lp) {
        try {
            this.mNetResources = this.mContext.getPackageManager().getResourcesForApplication(lp.getLaunchIntent().getComponent().getPackageName());
        } catch (NameNotFoundException e) {
            e.printStackTrace();
        }
        this.mNetResourcesSet = true;
    }

    private Drawable getNetResourceDrawable(LaunchPoint launchPoint, String resName) {
        int resId = 0;
        if (this.mNetResources != null) {
            resId = this.mNetResources.getIdentifier(resName, "drawable", launchPoint.getPackageName());
        }
        if (resId != 0) {
            Drawable ret = this.mNetResources.getDrawable(resId, null);
            if (ret != null) {
                return ret;
            }
        }
        Resources launcherRes = this.mContext.getResources();
        return launcherRes.getDrawable(launcherRes.getIdentifier(resName, "drawable", this.mContext.getPackageName()), null);
    }

    public void onSettingsChanged() {
        refreshDataSetAsync();
    }

    public void onLaunchPointsAddedOrUpdated(ArrayList<LaunchPoint> arrayList) {
    }

    public void onLaunchPointsRemoved(ArrayList<LaunchPoint> arrayList) {
    }
}

File: app/src/main/java/com/amazon/tv/leanbacklauncher/apps/SettingsAdapter.kt
package com.amazon.tv.leanbacklauncher.apps

import android.content.Context
import android.content.pm.PackageManager
import android.content.res.Resources
import android.graphics.drawable.Drawable
import android.os.Handler
import android.os.Message
import com.amazon.tv.firetv.leanbacklauncher.apps.AppCategory
import com.amazon.tv.leanbacklauncher.R
import java.lang.ref.WeakReference
import java.util.*

class SettingsAdapter(context: Context?, private val mConnectivityListener: ConnectivityListener) : AppsAdapter(context!!, null, AppCategory.SETTINGS) {
    private val mHandler: Handler = NetworkUpdateHandler(this)
    private var mNetResources: Resources? = null
    private var mNetResourcesSet = false

    private class NetworkUpdateHandler(adapter: SettingsAdapter?) : Handler() {
        private val mAdapterRef: WeakReference<SettingsAdapter?>
        override fun handleMessage(msg: Message) {
            val adapter = mAdapterRef.get()
            adapter?.let {
                when (msg.what) {
                    1 -> {
                        val index = it.updateNetwork()
                        if (index >= 0) {
                            it.notifyItemChanged(index)
                            return
                        }
                        return
                    }
                    else -> return
                }
            }
        }

        init {
            mAdapterRef = WeakReference<SettingsAdapter?>(adapter)
        }
    }

    private inner class SettingsComparator : Comparator<LaunchPoint> {
        override fun compare(lhs: LaunchPoint, rhs: LaunchPoint): Int {
            if (lhs.priority != rhs.priority) {
                return rhs.priority - lhs.priority
            }
            if (lhs.title == null) {
                return -1
            }
            return if (rhs.title == null) {
                1
            } else lhs.title!!.compareTo(rhs.title!!, ignoreCase = true)
        }
    }

    override fun onPostRefresh() {
        mNetResourcesSet = false
        updateNetwork()
    }

    fun onConnectivityChange() {
        mHandler.sendEmptyMessage(1)
    }

    private fun updateNetwork(): Int {
        for (i in mLaunchPoints.indices) {
            val launchPoint = mLaunchPoints[i]
            if (launchPoint?.settingsType == 0) {
                setNetwork(mContext.resources, launchPoint)
                return i
            }
        }
        return -1
    }

    private fun setNetwork(res: Resources, launchPoint: LaunchPoint): LaunchPoint {
        if (!mNetResourcesSet) {
            setNetworkResources(launchPoint)
        }
        val connectivityStatus = mConnectivityListener.connectivityStatus
        var titleId = 0
        var title: String? = null
        var hasNetworkName = false
        launchPoint.iconDrawable = null
        when (connectivityStatus.mNetworkType) {
            1 -> {
                titleId = R.string.settings_network
                launchPoint.iconDrawable = getNetResourceDrawable(launchPoint, "network_state_disconnected")
            }
            3, 5 -> {
                title = connectivityStatus.mWifiSsid
                hasNetworkName = true
                when (connectivityStatus.mWifiSignalStrength) {
                    0 -> launchPoint.iconDrawable = getNetResourceDrawable(launchPoint, "network_state_wifi_0")
                    1 -> launchPoint.iconDrawable = getNetResourceDrawable(launchPoint, "network_state_wifi_1")
                    2 -> launchPoint.iconDrawable = getNetResourceDrawable(launchPoint, "network_state_wifi_2")
                    3 -> launchPoint.iconDrawable = getNetResourceDrawable(launchPoint, "network_state_wifi_3")
                    4 -> launchPoint.iconDrawable = getNetResourceDrawable(launchPoint, "network_state_wifi_4")
                    else -> {
                    }
                }
            }
            7 -> {
                launchPoint.iconDrawable = getNetResourceDrawable(launchPoint, "network_state_ethernet")
                titleId = R.string.settings_network
            }
            9 -> {
                title = connectivityStatus.mWifiSsid
                hasNetworkName = true
                launchPoint.iconDrawable = getNetResourceDrawable(launchPoint, "network_state_wifi_no_internet")
            }
            11 -> {
                titleId = R.string.settings_network
                launchPoint.iconDrawable = getNetResourceDrawable(launchPoint, "network_state_ethernet_no_internet")
            }
            13 -> {
                title = connectivityStatus.mMobileNetworkName
                hasNetworkName = true
                when (connectivityStatus.mMobileSignalStrength) {
                    0 -> launchPoint.iconDrawable = getNetResourceDrawable(launchPoint, "network_state_cellular_0")
                    1 -> launchPoint.iconDrawable = getNetResourceDrawable(launchPoint, "network_state_cellular_1")
                    2 -> launchPoint.iconDrawable = getNetResourceDrawable(launchPoint, "network_state_cellular_2")
                    3 -> launchPoint.iconDrawable = getNetResourceDrawable(launchPoint, "network_state_cellular_3")
                    4 -> launchPoint.iconDrawable = getNetResourceDrawable(launchPoint, "network_state_cellular_4")
                    else -> {
                    }
                }
            }
            15 -> {
                title = connectivityStatus.mMobileNetworkName
                hasNetworkName = true
                when (connectivityStatus.mMobileSignalStrength) {
                    0 -> launchPoint.iconDrawable = getNetResourceDrawable(launchPoint, "network_state_cellular_no_internet_0")
                    1 -> launchPoint.iconDrawable = getNetResourceDrawable(launchPoint, "network_state_cellular_no_internet_1")
                    2 -> launchPoint.iconDrawable = getNetResourceDrawable(launchPoint, "network_state_cellular_no_internet_2")
                    3 -> launchPoint.iconDrawable = getNetResourceDrawable(launchPoint, "network_state_cellular_no_internet_3")
                    4 -> launchPoint.iconDrawable = getNetResourceDrawable(launchPoint, "network_state_cellular_no_internet_4")
                    else -> {
                    }
                }
            }
        }
        if (launchPoint.iconDrawable == null) {
            launchPoint.iconDrawable = getNetResourceDrawable(launchPoint, "network_state_wifi_0")
        }
        if (titleId != 0) {
            title = res.getString(titleId)
        }
        if (title == null) {
            title = res.getString(R.string.settings_network)
        }
        launchPoint.title = title
        if (hasNetworkName) {
            launchPoint.contentDescription = res.getString(R.string.settings_network)
        }
        return launchPoint
    }

    override fun onRankerReady() {}
    override fun sortLaunchPoints(launchPoints: ArrayList<LaunchPoint>) {
        Collections.sort(launchPoints, SettingsComparator())
    }

    private fun setNetworkResources(lp: LaunchPoint) {
        try {
            mNetResources = mContext.packageManager.getResourcesForApplication(lp.launchIntent!!.component!!.packageName)
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
        }
        mNetResourcesSet = true
    }

    private fun getNetResourceDrawable(launchPoint: LaunchPoint, resName: String): Drawable {
        var resId = 0
        if (mNetResources != null) {
            resId = mNetResources!!.getIdentifier(resName, "drawable", launchPoint.packageName)
        }
        if (resId != 0) {
            val ret = mNetResources!!.getDrawable(resId, null)
            if (ret != null) {
                return ret
            }
        }
        val launcherRes = mContext.resources
        return launcherRes.getDrawable(launcherRes.getIdentifier(resName, "drawable", mContext.packageName), null)
    }

    override fun onSettingsChanged() {
        refreshDataSetAsync()
    }

//    override fun onLaunchPointsAddedOrUpdated(arrayList: ArrayList<LaunchPoint>) {}

//    override fun onLaunchPointsRemoved(arrayList: ArrayList<LaunchPoint>) {}

    companion object {
        private const val TAG = "SettingsAdapter"
    }
}